### PR TITLE
HOTFIX: fix potentially hanging test shouldAddStateStoreToRegexDefine…

### PR DESCRIPTION
**Pull request by bbejeck**
…dSource

### CODE REVIEW

**Comment by bbejeck** [comment]
@guozhangwang @mjsax 

Updated the test in question.  I couldn't reproduce locally, but that could be because I'm running on a  Mac (10.10.5).  But at any rate, it's consistent now with other tests that have had no issues.

File: streams/src/test/java/org/apache/kafka/streams/integration/RegexSourceIntegrationTest.java [hunk 1]
```diff
@@ -251,12 +251,14 @@ public void shouldAddStateStoreToRegexDefinedSource() throws Exception {
         streams.start();
 
         final Properties producerConfig = TestUtils.producerConfig(CLUSTER.bootstrapServers(), StringSerializer.class, StringSerializer.class);
+        final Properties consumerConfig = TestUtils.consumerConfig(CLUSTER.bootstrapServers(), StringDeserializer.class, StringDeserializer.class);
+
+        IntegrationTestUtils.produceValuesSynchronously(TOPIC_1, Arrays.asList("testMessage"), producerConfig, mockTime);
+        IntegrationTestUtils.waitUntilMinKeyValueRecordsReceived(consumerConfig, TOPIC_1, 1, 5000);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/RegexSourceIntegrationTest.java]: Please increase timeout to 30 seconds.

**Review by mjsax — CHANGES_REQUESTED** [review]
Thanks. One minor comment.

**Comment by bbejeck** [comment]
@mjsax thanks for the review, timeout increased.

**Comment by mjsax** [comment]
Retest this please.

**Review by mjsax — APPROVED** [review]
LGTM.

Call for second review \cc @enothereska @dguy (the test is question did hang once and did timeout the build)

**Comment by enothereska** [comment]
LGTM

**Comment by mjsax** [comment]
Call for final review and merging @ijuma

**Comment by guozhangwang** [comment]
retest this please.

**Comment by guozhangwang** [comment]
retest this please.

**Comment by guozhangwang** [comment]
retest this please

**Comment by mjsax** [comment]
I guess it might be helpful to rebase this to get 3 green builds -- we got a lot of test stabilization fixes recently... WDYT?

**Comment by bbejeck** [comment]
@mjsax  agreed, I'll rebase this PR as soon as I have access to my machine.

**Comment by bbejeck** [comment]
rebased - let's see if this fixes the tests

**Comment by bbejeck** [comment]
retest this please.

**Comment by bbejeck** [comment]
retest this please.

**Comment by bbejeck** [comment]
retest this please.

**Comment by bbejeck** [comment]
looks like this PR is still having issues, I'll rebase again and see if that helps

**Comment by bbejeck** [comment]
@guozhangwang @mjsax  finally 3 greens

File: streams/src/test/java/org/apache/kafka/streams/integration/RegexSourceIntegrationTest.java [hunk 2]
```diff
@@ -254,12 +254,14 @@ public void shouldAddStateStoreToRegexDefinedSource() throws Exception {
         streams.start();
 
         final Properties producerConfig = TestUtils.producerConfig(CLUSTER.bootstrapServers(), StringSerializer.class, StringSerializer.class);
+        final Properties consumerConfig = TestUtils.consumerConfig(CLUSTER.bootstrapServers(), StringDeserializer.class, StringDeserializer.class);
+
+        IntegrationTestUtils.produceValuesSynchronously(TOPIC_1, Arrays.asList("testMessage"), producerConfig, mockTime);
+        IntegrationTestUtils.waitUntilMinKeyValueRecordsReceived(consumerConfig, TOPIC_1, 1, 30000);
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/integration/RegexSourceIntegrationTest.java]: nit: we can just call the function with three params which will then use `DEFAULT_TIMEOUT = 30 * 1000L;`.

**Review by guozhangwang — COMMENTED** [review]
@bbejeck lgtm. Just a minor comment.

**Comment by bbejeck** [comment]
@guozhangwang  comment addressed.

**Comment by guozhangwang** [comment]
@bbejeck After thinking a bit more about it I think there is still a small risk, because a separate consumer being able to consume the message (meaning that the consumer has discovered the topic and hence being able to fetch from it) does not necessarily mean that the streams' embedded consumer, which is independent, would be able to discover the new topic in time as well. Though adding this one more line does reduce the timing issue by a lot. 

So I think a safer fix would be 0) do not close Streams right after producing the data, 1) use `TestUtils.waitForCondition` to continuously check that `stateStoreToSourceTopic = builder.stateStoreNameToSourceTopics()`, until say 30 seconds has elapsed and then report the test as failed. 2) call `streams.close()` in a finally block.

WDYT?

**Comment by bbejeck** [comment]
@guozhangwang  yeah I think that is a better approach, thanks for the suggestion.  I'll make the corresponding updates.

**Comment by bbejeck** [comment]
@guozhangwang made changes per comments. 

Also, looking at the test some more, I removed the part of publishing and consuming a message.  IMHO the test is really about making sure that we associate state stores with the discovered topics matching the regex.  That is once `StreamPartitionsAssignor` calls `builder.updateSubscriptions`, the right actions take place.   WDYT?

**Comment by guozhangwang** [comment]
@bbejeck totally agree!

**Review by guozhangwang — APPROVED** [review]
Merged to trunk.

**Pull request closed without merging**